system/src/system_network_compat.cpp

```diff
@@ -147,7 +148,24 @@ void network_off(network_handle_t network, uint32_t flags, uint32_t param, void*
 }
 
 int network_wait_off(network_handle_t network, system_tick_t timeout, void*) {
-    // Temporarily not necessary for Gen2 platforms.
+    /* This is a workaround for the power leakage issue that happens when the
+     * device is powered on and it enters sleep before the modem is initialized.
+     */
+#if PLATFORM_ID == PLATFORM_ELECTRON
+    cellular_pause(nullptr);
+    if (HAL_GPIO_Read(RXD_UC)) {
```

Comment:
I know we've discussed this, but now that I've taken a look at the modem code in the context of #2198, it should probably be fine to use `RI_UC` as long as `AT+URING` was not used.

I propose we should also try to hide as much of the platform-specific behavior as possible. Can we simply modify https://github.com/particle-iot/device-os/blob/develop/hal/src/electron/modem/mdm_hal.cpp#L1114 to perhaps take a look at the current `RI_UC` or some other state and perform the power-off sequence anyway? It should also be fine to introduce an additional cellular hal method if required.

Reply:
I'd like to refactor it after #2198 being merged to reduce conflicts on rebasing.

Reply:
You can base this PR on that branch for now.

Reply:
It's not just about how we perform the power-off sequence according to the current modem hal, but also the `SPARK_WLAN_STARTED` in the `off()` method in system_network_internal.h. When the power leakage encountered, the `SPARK_WLAN_STARTED` isn't set. Besides, this `off()` implementation is also shared with Photon/P1. I feel like if we're going to introduce a new HAL API for managing the modem power states, that is another approach we mentioned, which will refactor a bunch of files.

Reply:
If `SPARK_WLAN_STARTED` isn't set, we can detect that in the system layer and still perhaps perform `off_now()` there? Where the HAL can take care of checking the current state and turning off if needed.

Reply:
We can also initialize to by default to something like `UNKNOWN` state and act on it accordingly.

Reply:
Its type is `uint8_t`, and any other value except 0 and 1 will be `true` to perform some logic behavior accordingly for those where `if (SPARK_WLAN_STARTED)` is used.

Reply:
It's feasible though, we just need to be careful of where `SPARK_WLAN_STARTED` is referenced.

Reply:
>If SPARK_WLAN_STARTED isn't set, we can detect that in the system layer and still perhaps perform off_now() there?

If the modem is initialized and powered off by user application, `SPARK_WLAN_STARTED` is cleared as well, while in which case we don't need to perform that particular power-off sequence.

Reply:
I mean if we want to be careful with `SPARK_WLAN_STARTED`, okay, we can introduce another flag that would tell us that we are in an unknown state perhaps? :)

Reply:
Not bad. We have anyway introduced several global flags for managing the network status for Gen2 already. It can't be worse.